source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc

```diff
@@ -72,90 +175,127 @@ void Filter::onReadWorker() {
           "[" + std::string{line_parts[3]} + "]:" + std::string{line_parts[5]});
     } else {
       throw EnvoyException("failed to read proxy protocol");
     }
+  }
+}
 
-    // Error check the source and destination fields. Most errors are caught by the address
-    // parsing above, but a malformed IPv6 address may combine with a malformed port and parse as
-    // an IPv6 address when parsing for an IPv4 address. Remote address refers to the source
-    // address.
-    const auto remote_version = remote_address->ip()->version();
-    const auto local_version = local_address->ip()->version();
-    if (remote_version != protocol_version || local_version != protocol_version) {
-      throw EnvoyException("failed to read proxy protocol");
-    }
-    // Check that both addresses are valid unicast addresses, as required for TCP
-    if (!remote_address->ip()->isUnicastAddress() || !local_address->ip()->isUnicastAddress()) {
-      throw EnvoyException("failed to read proxy protocol");
+bool Filter::readProxyHeader(int fd, WireHeader& hdr) {
+  while (buf_off_ < MAX_PROXY_PROTO_LEN) {
+    int bytes_avail;
+
+    if (ioctl(fd, FIONREAD, &bytes_avail) < 0) {
+      throw EnvoyException("failed to read proxy protocol (no bytes avail)");
     }
 
-    // Only set the local address if it really changed, and mark it as address being restored.
-    if (*local_address != *socket.localAddress()) {
-      socket.setLocalAddress(local_address, true);
+    if (bytes_avail == 0) {
+      return false;
     }
-    socket.setRemoteAddress(remote_address);
-  }
 
-  // Release the file event so that we do not interfere with the connection read events.
-  file_event_.reset();
-  cb_->continueFilterChain(true);
-}
+    bytes_avail = std::min(size_t(bytes_avail), MAX_PROXY_PROTO_LEN - buf_off_);
 
-bool Filter::readLine(int fd, std::string& s) {
-  while (buf_off_ < MAX_PROXY_PROTO_LEN) {
-    ssize_t nread = recv(fd, buf_ + buf_off_, MAX_PROXY_PROTO_LEN - buf_off_, MSG_PEEK);
+    ssize_t nread = recv(fd, buf_ + buf_off_, bytes_avail, MSG_PEEK);
 
     if (nread == -1 && errno == EAGAIN) {
       return false;
     } else if (nread < 1) {
-      throw EnvoyException("failed to read proxy protocol");
+      throw EnvoyException("failed to read proxy protocol (no bytes read)");
     }
 
-    bool found = false;
-    // continue searching buf_ from where we left off
-    for (; search_index_ < buf_off_ + nread; search_index_++) {
-      if (buf_[search_index_] == '\n' && buf_[search_index_ - 1] == '\r') {
-        search_index_++;
-        found = true;
-        break;
+    if (buf_off_ + nread >= PROXY_PROTO_V2_HEADER_LEN) {
+      const char* sig = PROXY_PROTO_V2_SIGNATURE;
+      if (!memcmp(buf_, sig, PROXY_PROTO_V2_SIGNATURE_LEN)) {
+        header_version_ = V2;
+      } else if (memcmp(buf_, PROXY_PROTO_V1_SIGNATURE, PROXY_PROTO_V1_SIGNATURE_LEN)) {
+        // If its not v2, and can't be v1
```

Comment:
please reflow to 100 cols and use proper punctuation, etc. if possible. Same elsewhere.